src/go/interests/interests_test.go

```diff
@@ -0,0 +1,129 @@
+//  Copyright 2018 Google LLC
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+
+//        https://www.apache.org/licenses/LICENSE-2.0
+
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//	limitations under the License.
+
+package interests
+
+import (
+	"errors"
+	"testing"
+
+	mpb "github.com/google/minions/proto/minions"
+	opb "github.com/google/minions/proto/overlord"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMinifyOnSamePathKeepsMetadataAndData(t *testing.T) {
+	i1 := &mpb.Interest{
+		DataType:   mpb.Interest_METADATA_AND_DATA,
+		PathRegexp: "/some/regexp",
+	}
+	i2 := &mpb.Interest{
+		DataType:   mpb.Interest_METADATA,
+		PathRegexp: "/some/regexp",
+	}
+	interests := []*mpb.Interest{i1, i2}
+	minified := Minify(interests)
+	require.Len(t, minified, 1)
+	require.Equal(t, minified[0], i1)
+}
+
+func TestMinifyOnDifferentPathDoesNotDedupe(t *testing.T) {
+	i1 := &mpb.Interest{
+		DataType:   mpb.Interest_METADATA_AND_DATA,
+		PathRegexp: "/some/regexp",
+	}
+	i2 := &mpb.Interest{
+		DataType:   mpb.Interest_METADATA_AND_DATA,
+		PathRegexp: "/some/regexp2",
+	}
+	interests := []*mpb.Interest{i1, i2}
+	minified := Minify(interests)
+	require.Contains(t, minified, i1)
+	require.Contains(t, minified, i2)
+}
+
+var isMatchingV = []struct {
+	title string
+	i     *mpb.Interest
+	f     *opb.File
+	e     error
+	r     bool
+}{
+	{"empty_interest_file", &mpb.Interest{}, &opb.File{}, errors.New(""), false},
+	{"no_file_metadata", &mpb.Interest{
+		DataType:   mpb.Interest_METADATA_AND_DATA,
+		PathRegexp: "/",
+	}, &opb.File{}, errors.New(""), false},
```

Comment:
For future reference, no need to address this now:
If you move [this error](https://github.com/google/minions/blob/44140250b943f30030b898ae5d63f8c248edbb08/src/go/overlord/interests/interests.go#L57) out of the function and declare it on a package level, you might be able to distinguish between "an error occurred" and "my error occurred", so that you can see when a regex goes wrong or when getting metadata is broken during testing.

Reply:
I actually wondered extensively about that, thanks for pointing it out.
And by "distinguishing" effectively you mean "doing a string comparison of the error object' right, possibly adding one of those bizarre "isThisMyError(error)" methods, not actually creating its own error struct?

Reply:
Never do string comparison on errors please.

You would have 
```go
var errNoMetadata = fmt.Errorf("cannot match file without metadata")
```
On a package level, then [here](https://github.com/google/minions/blob/44140250b943f30030b898ae5d63f8c248edbb08/src/go/overlord/interests/interests.go#L57) change it to `return false, errNoMetadata`.

After that you can check the error with `err == errNoMetadata`.

Since this is not exported you don't need to create helper methods. If you are going to add more information like format strings, additional values etc, please create a struct of your type, which is the equivalent of creating a new class that extends Exception in Java.

Java:
```java
class MyException extends Exception
{
      public MyException(String message)
      {
         super(message);
      }
 }
// [...]
try
{
          brokenStuff(); // throw new MyException();
}
catch(MyException ex)
{
      // Process
}
```
Go:
```go
type MyErr string
func(m *MyErr) Error() string { return string(m) }
// [...]
err := brokenStuff();
if me, ok := err.(MyErr); ok {
// Process
}
```

Alternatively if you have multiple types use a switch (which behaves exactly as a `catch` block)
```go
switch err.(type) {
case MyErr:
// Process
case MyOtherErr:
// Process
}
```

This can be extended to interfaces, for example if you only care about the "temporary" quality about an error (as in: "should you retry?")

```go
type myTempErr1 struct{}
func(myTempErr1) WaitBeforeRetry() time.Duration { return time.Second }

type TempErr interface {
     WaitBeforeRetry() time.Duration
}

// [...]
for {
    err := temporarilyBrokenStuff()

    if tempErr, ok := err.(TempErr); ok {
         time.Sleep(tempErr.WaitBeforeRetry())
    } else {
         return err
    }
}
```
And you can have several errors implementing that interface to do the checking.

Reply:
Ah, this was very instructive, thank you. I'm already slightly less annoyed by golang err representation.
Let me do some more changes then.

Reply:
I take it back - now that I've implemented the change I've already run into the ubercommon situation that some of the most important libraries (e.g. regexp) do not provide any type whatsoever for the errors. So I can't really do anything useful in terms of distinguishing the error, except saying "my stuff" or "anything else"

I think it's OK in this context as it is. Of course I could go and check the code to see what's being thrown, but I don't think that's some great design at work.